Grafana notifications: Fix panel/dashboardId type

bpedersen2 opened the pull request:
<!--- changelog-entry
# Fill in 'msg' below to have an entry automatically added to the next release changelog.
# Leaving 'msg' blank will not generate a changelog entry for this PR.
# Please ensure this is a simple (and readable) one-line string.
---
msg: "Ensure grafana IDs are integers in notifications "
-->

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Latest grafana fails with
  Error sending notification grafana: 400
  [{"classification":"DeserializationError",
    "message":"json: cannot unmarshal string into Go struct
        field PostAnnotationsCmd.dashboardId of type int64"}]

So ensure the IDs are really inst and not strings.

<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - API

##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
awx: 19.3.0

```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

Try sending notifcations to grafana v8.1.3 to reproduce this bug.

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

bpedersen2 committed:
Grafana notifications: Fix panel/dashboardId type

Latest grafana fails with
  Error sending notification grafana: 400
  [{"classification":"DeserializationError",
    "message":"json: cannot unmarshal string into Go struct
        field PostAnnotationsCmd.dashboardId of type int64"}]

So ensure the IDs are really int and not strings.

francoisv100 commented:
Hello 

Tested suggested python code changes to the file awx/main/notifications/grafana_backend.py and the suggested changes work, not sure how to proceed from here to get the changes merged into awx.

Tested grafana notifications to Grafana version 8.2.5.

![Grafana Notification Dashboard](https://user-images.githubusercontent.com/95674398/151569332-cc97efc8-58e6-4e42-837a-02bcb51fc9b9.png)

![Grafana Notification Test JSON output](https://user-images.githubusercontent.com/95674398/151569348-40786ca3-311a-4d1d-bf70-072cdefe2b0c.png)

<img width="1148" alt="Grafana Notification Test working 1" src="https://user-images.githubusercontent.com/95674398/151570088-8f43ac08-704e-4a81-bd61-1c014fed6eff.png">

bpedersen2 commented:
Ping...

kdelee commented:
Looks like you need to update `awx/main/tests/unit/notifications/test_grafana.py` to deal with how somtimes you might not have a `panelId` etc

bpedersen2 committed:
Fix the dashboard/panelId=0 case

0 is avlaid valid for the ID's, so ensure to allow them.

bpedersen2 committed:
Update tests to new behavior

Panel/Dashboard Id fields are not sent if they where not requested.
Alos add tests for the ID=0 case.

bpedersen2 commented:
> Looks like you need to update `awx/main/tests/unit/notifications/test_grafana.py` to deal with how somtimes you might not have a `panelId` etc

Done

francoisv100 commented:
tested the below scenarios without editing awx/main/tests/unit/notifications/test_grafana.py , not sure how relevant this is testing from the web frontend.

only modified , awx/main/notifications/grafana_backend.py on local test machine,

tested notifications with panelId and no dashboardId
tested notifications with no panelId and dashboardId
tested notifications with no panelId and no dashboardId
tested notifications with no panelId and no dashboardId and no tags

all 4 scenarios work from AWX web frontend,

Grafana /api/annotations :

[{"id":11134,"alertId":0,"alertName":"","dashboardId":0,"panelId":58,"userId":0,"newState":"","prevState":"","created":1644474757957,"updated":1644474757957,"time":1644474757957,"timeEnd":1644474757957,"text":"Notification Test 1 https://towerhost","tags":["AWX ALARMS"],"login":"","email":"","avatarUrl":"","data":{}},

{"id":11133,"alertId":0,"alertName":"","dashboardId":3,"panelId":0,"userId":0,"newState":"","prevState":"","created":1644474708182,"updated":1644474708182,"time":1644474708182,"timeEnd":1644474708182,"text":"Notification Test 1 https://towerhost","tags":["AWX ALARMS"],"login":"","email":"","avatarUrl":"","data":{}},

{"id":11132,"alertId":0,"alertName":"","dashboardId":0,"panelId":0,"userId":0,"newState":"","prevState":"","created":1644474584248,"updated":1644474584248,"time":1644474584248,"timeEnd":1644474584248,"text":"Notification Test 1 https://towerhost","tags":["AWX ALARMS"],"login":"","email":"","avatarUrl":"","data":{}},

[{"id":11137,"alertId":0,"alertName":"","dashboardId":0,"panelId":0,"userId":0,"newState":"","prevState":"","created":1644476096351,"updated":1644476096351,"time":1644476096351,"timeEnd":1644476096351,"text":"Notification Test 1 https://towerhost","tags":[],"login":"","email":"","avatarUrl":"","data":{}}

AWX logs:

tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 GET / - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 GET /api/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 GET /api/v2/auth/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 GET /api/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 GET /api/login/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 2022-02-10 06:28:10,810 INFO     [295f0848] awx.api.generics User root logged in from 192.168.10.39
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 POST /api/login/ - HTTP/1.1 302
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 GET /api/v2/config/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 GET /api/v2/config/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 GET /api/v2/me/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 GET /api/v2/users/2/admin_of_organizations/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 GET /api/v2/organizations/?page_size=1&role_level=notification_admin_role - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 GET /api/v2/organizations/?page_size=1&role_level=execution_environment_admin_role - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 GET /api/v2/workflow_approvals/?page_size=1&status=pending - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 GET /api/v2/dashboard/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 GET /api/v2/dashboard/graphs/jobs/?job_type=all&period=month - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 GET /api/v2/dashboard/graphs/jobs/?job_type=all&period=month - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 GET /api/v2/notification_templates/?order_by=name&page=1&page_size=20 - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 OPTIONS /api/v2/notification_templates/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 OPTIONS /api/v2/notification_templates/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 GET /api/v2/notification_templates/1/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 GET /api/v2/organizations/?order_by=name&page=1&page_size=5 - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 OPTIONS /api/v2/organizations/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 PATCH /api/v2/notification_templates/1/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 GET /api/v2/notification_templates/1/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 OPTIONS /api/v2/notification_templates/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 GET /api/v2/notification_templates/1/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 OPTIONS /api/v2/notification_templates/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 OPTIONS /api/v2/organizations/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 GET /api/v2/organizations/?order_by=name&page=1&page_size=5 - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 GET /api/v2/organizations/?order_by=name&page=1&page_size=5 - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 OPTIONS /api/v2/organizations/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 PATCH /api/v2/notification_templates/1/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 GET /api/v2/notification_templates/1/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 OPTIONS /api/v2/notification_templates/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 POST /api/v2/notification_templates/1/test/ - HTTP/1.1 202
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 GET /api/v2/notifications/10/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 GET /api/v2/notification_templates/1/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 OPTIONS /api/v2/notification_templates/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 GET /api/v2/organizations/?order_by=name&page=1&page_size=5 - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 OPTIONS /api/v2/organizations/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 GET /api/v2/organizations/?order_by=name&page=1&page_size=5 - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 OPTIONS /api/v2/organizations/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 PATCH /api/v2/notification_templates/1/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 GET /api/v2/notification_templates/1/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 OPTIONS /api/v2/notification_templates/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 POST /api/v2/notification_templates/1/test/ - HTTP/1.1 202
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 GET /api/v2/notifications/11/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 OPTIONS /api/v2/notification_templates/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 GET /api/v2/organizations/?order_by=name&page=1&page_size=5 - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 OPTIONS /api/v2/organizations/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 GET /api/v2/notification_templates/1/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 GET /api/v2/organizations/?order_by=name&page=1&page_size=5 - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 OPTIONS /api/v2/organizations/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 PATCH /api/v2/notification_templates/1/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 GET /api/v2/notification_templates/1/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 OPTIONS /api/v2/notification_templates/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 POST /api/v2/notification_templates/1/test/ - HTTP/1.1 202
tools_awx_1     | awx-uwsgi stdout | 
tools_awx_1     | awx-uwsgi stdout | 192.168.10.39 GET /api/v2/notifications/12/ - HTTP/1.1 200
tools_awx_1     | awx-uwsgi stdout |

francoisv100 commented:
Hello Elijah Delee 

Please run test again.

francoisv100 commented:
Hello Elijah Delee

Please let us know when the required tests/checks can be executed again for this PR ?

francoisv100 commented:
Hello Everyone 

Any new update here on when the workflow approval/tests can be run for the PR to merge ?

bpedersen2 commented:
thetest failures seem unrelated...

francoisv100 commented:
Hello

Thanks, How do we proceed then to get the PR merged ? , only option I see here other than the tests is for project collaborators to action the merge request.

francoisv100 commented:
Hello Any new feedback on how to proceed ?

rebeccahhh reviewed (approved):
I've got CI passing now and the scope of this is small, I've got some minor testing on my end I'm waiting on and then I'll merge. thanks for contributing :)

francoisv100 commented:
Thanks Rebecca.
